src/gurt/shm_lru.c

```diff
@@ -0,0 +1,475 @@
+				node_hnext->off_hnext = node->off_hnext;
+				node->off_hnext       = 0;
+				break;
+			}
+			offset = node_hnext->off_hnext;
+		}
+	}
+
+	first_av_saved = subcache->first_av;
+	subcache->first_av =
+	    (off_node_to_remove - (long int)subcache->off_nodelist) / sizeof(shm_lru_node_t);
+
+	/* remove from list */
+	if (node->off_prev > 0) {
+		node_prev = (shm_lru_node_t *)((long int)cache + (long int)node->off_prev);
+		node_prev->off_next = node->off_next;
+	}
+	if (node->off_next > 0) {
+		node_next = (shm_lru_node_t *)((long int)cache + (long int)node->off_next);
+		node_next->off_prev = node->off_prev;
+	}
+
+	if (node == node_tail)
+		subcache->off_tail = node->off_prev;
+
+	if (off_node_to_remove == subcache->off_head)
+		subcache->off_head = node->off_next;
+
+	node->off_hnext = first_av_saved;
+	subcache->size--;
+	if (subcache->size == 0)
+		/* cache is empty */
+		subcache->off_head = 0;
+	return 0;
+}
+
+static inline int
+key_cmp(shm_lru_cache_t *cache, shm_lru_node_t *node, void *key, uint32_t key_size)
+{
+	if (node->key_size != key_size)
+		/* key size does not match */
+		return 1;
+
+	return memcmp((char *)cache + node->key, key, key_size);
+}
+
+int
+shm_lru_get(shm_lru_cache_t *cache, void *key, uint32_t key_size, shm_lru_node_t **node_found,
+	    void **val)
+{
+	uint64_t             hash         = d_hash_murmur64(key, key_size, 0);
+	uint32_t             idx_subcache = (uint32_t)((hash >> 32) % cache->n_subcache);
+	shm_lru_cache_var_t *cache_list =
+	    (shm_lru_cache_var_t *)((long int)cache + cache->off_cache_list);
+	int *off_hash =
+	    (int *)((long int)cache + (long int)cache_list[idx_subcache].off_hashbuckets);
+	uint32_t        index  = (uint32_t)(hash % cache->capacity);
+	int             offset = off_hash[index];
+	shm_lru_node_t *node;
+
+	shm_mutex_lock(&cache_list[idx_subcache].lock);
+
+	*node_found = NULL;
+	if (cache->key_size != 0)
+		D_ASSERT(cache->key_size == key_size);
+
+	while (offset) {
+		node = (shm_lru_node_t *)((long int)cache + (long int)offset);
+		if (key_cmp(cache, node, key, key_size) == 0) {
+			lru_move_to_head(cache, &cache_list[idx_subcache], node);
+			*val = ((void *)cache + (long int)node->data);
+			atomic_fetch_add(&node->ref_count, 1);
+			*node_found = node;
+			shm_mutex_unlock(&cache_list[idx_subcache].lock);
+			return 0;
+		}
+		if (node->off_hnext == 0) {
+			shm_mutex_unlock(&cache_list[idx_subcache].lock);
+			return ENOENT;
+		}
+		offset = node->off_hnext;
+	}
+	shm_mutex_unlock(&cache_list[idx_subcache].lock);
+
+	return ENOENT;
+}
+
+/* Put key-data into cache */
+int
+shm_lru_put(shm_lru_cache_t *cache, void *key, uint32_t key_size, void *data, uint32_t data_size)
+{
+	int                  rc;
+	uint64_t             hash         = d_hash_murmur64(key, key_size, 0);
+	uint32_t             idx_subcache = (uint32_t)((hash >> 32) % cache->n_subcache);
+	uint32_t             index        = (uint32_t)(hash % cache->capacity);
```

Comment:
do you have some data that  show this splitting of high 32 to subcache and low 32 to index in subcache is optimal?

Reply:
Code was updated to pick sub-cache and bucket in sub-cache using 64-bit hash.
I did check the distribution of 1) only using high 32bits of the hash 2) only using low 32bits of the hash 3) using 64 bits of the hash. The differences are indistinguishable.